Add RFC on OpenTelemetry (#1757)
openbao/openbao: website/content/docs/rfcs/opentelemetry.mdx

# Switch to OpenTelemetry

## Summary

> Should we drop [`go-metrics`][] library and thereby support for the metrics

Reviewer: This is a level of sorcery I've not yet seen :laughing: Didn't realize you could leave the link blank and have it auto-fill to the text value.

# Switch to OpenTelemetry

## Summary

Should we drop [`go-metrics`][] library and thereby support for the metrics
protocols of:

- Circonus
- Datadog
- Stackdriver
- Statsd
- Statssite

in favor of [`opentelemetry-go`][]?

## Problem Statement

Supporting vendor specific protocols only really makes sense, if we actually
make use of their distinctive features. But as we are supporting them through a
generic API, we don't do this. The downside would be, that this is breaking
change.

> It is probably save to say that OpenTelemetry won the format war on telemetry

Reviewer 2: ```suggestion
It is probably safe to say that OpenTelemetry won the format war on telemetry
```

(at least for the time being, until the next big thing comes). At some point
most - if not all - metrics consumers/sink will either support it or become
irrelevant.

We could also use the chance to clean up the existing metrics:

- use labels instead of dynamic names ([as discussed on GitHub][metric names
  discussion])
- define an API for plugins to expose metrics, with proper namespacing
- etc.

We can also mark the new metrics as experimental at first, which would give us
the flexibility to iterate on them for a few releases while gathering feedback.

## User Migration Path

> We should first add the OpenTelemetry based metrics and deprecated the
> [`go-metrics`][] based metrics in one release and in a later release remove the
> old metrics.

Reviewer 2: ```suggestion
We should first add the OpenTelemetry based metrics and deprecate the
[`go-metrics`][] in one release and in later releases remove them all together.
```
